ttab/model_adaptation/rotta.py

```diff
@@ -0,0 +1,664 @@
+            var = (1 - self.momentum) * self.source_var + self.momentum * b_var
+            self.source_mean, self.source_var = deepcopy(mean.detach()), deepcopy(
+                var.detach()
+            )
+            mean, var = mean.view(1, -1), var.view(1, -1)
+        else:
+            mean, var = self.source_mean.view(1, -1), self.source_var.view(1, -1)
+
+        x = (x - mean) / torch.sqrt(var + self.eps)
+        weight = self.weight.view(1, -1)
+        bias = self.bias.view(1, -1)
+
+        return x * weight + bias
+
+
+class RobustBN2d(MomentumBN):
+    def forward(self, x):
+        if self.training:
+            b_var, b_mean = torch.var_mean(
+                x, dim=[0, 2, 3], unbiased=False, keepdim=False
+            )  # (C,)
+            mean = (1 - self.momentum) * self.source_mean + self.momentum * b_mean
+            var = (1 - self.momentum) * self.source_var + self.momentum * b_var
+            self.source_mean, self.source_var = deepcopy(mean.detach()), deepcopy(
+                var.detach()
+            )
+            mean, var = mean.view(1, -1, 1, 1), var.view(1, -1, 1, 1)
+        else:
+            mean, var = self.source_mean.view(1, -1, 1, 1), self.source_var.view(
+                1, -1, 1, 1
+            )
+
+        x = (x - mean) / torch.sqrt(var + self.eps)
+        weight = self.weight.view(1, -1, 1, 1)
+        bias = self.bias.view(1, -1, 1, 1)
+
+        return x * weight + bias
+
+
+# ========== custom_transforms ==========
+def get_tta_transforms(cfg, gaussian_std: float = 0.005, soft=False):
+    img_shape = (*cfg.img_shape, 3)
+    n_pixels = img_shape[0]
+
+    clip_min, clip_max = 0.0, 1.0
+
+    p_hflip = 0.5
+
+    tta_transforms = transforms.Compose(
+        [
+            Clip(0.0, 1.0),
+            ColorJitterPro(
+                brightness=[0.8, 1.2] if soft else [0.6, 1.4],
+                contrast=[0.85, 1.15] if soft else [0.7, 1.3],
+                saturation=[0.75, 1.25] if soft else [0.5, 1.5],
+                hue=[-0.03, 0.03] if soft else [-0.06, 0.06],
+                gamma=[0.85, 1.15] if soft else [0.7, 1.3],
+            ),
+            transforms.Pad(padding=int(n_pixels / 2), padding_mode="edge"),
+            transforms.RandomAffine(
+                degrees=[-8, 8] if soft else [-15, 15],
+                translate=(1 / 16, 1 / 16),
+                scale=(0.95, 1.05) if soft else (0.9, 1.1),
+                shear=None,
+                interpolation=PIL.Image.BILINEAR,
+                fill=None,
+            ),
+            transforms.GaussianBlur(
+                kernel_size=5, sigma=[0.001, 0.25] if soft else [0.001, 0.5]
+            ),
+            transforms.CenterCrop(size=n_pixels),
+            transforms.RandomHorizontalFlip(p=p_hflip),
+            GaussianNoise(0, gaussian_std),
+            Clip(clip_min, clip_max),
+        ]
+    )
+    return tta_transforms
+
+
+class GaussianNoise(torch.nn.Module):
```

Comment:
This is irrelevant to this TTA method, right? So please move this to somewhere else or remove it.

Reply:
It is used in the `get_tta_transforms` function, as well as the functions below.

Reply:
I see. It seems these modules are transferable to other augmentation-based TTA methods. Then it makes sense to move these things to other places.